platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/http/LocalRequestTask.kt

```diff
@@ -0,0 +1,62 @@
+package org.maplibre.android.http
+
+import android.content.res.AssetManager
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.maplibre.android.MapLibre
+import org.maplibre.android.MapStrictMode
+import org.maplibre.android.log.Logger
+import java.io.IOException
+import java.io.InputStream
+
+internal class LocalRequestTask(private val scope: CoroutineScope, private val onCompletion: ((ByteArray?) -> Unit)?) {
+
+	fun execute(url: String) {
+		scope.launch(Dispatchers.IO) {
```

Comment:
(This comment applies to other Tasks, too)

This task has no state and is internal to MapLibre, i.e. potentially does not interface with Java code. The idiomatic Kotlin way for this would look like this, i.e. not a class but a suspending function)

```suggestion
suspend fun requestLocalUrl(url: String): ByteArray? = withContext(IO) {
   loadFile(...)
}
```

The current approach causes a leak of the coroutine scope because the `LocalRequestTask` is not garbage collected as long as the owner of `LocalRequestTask` is not. No idea how serious it is, my guess would be: not at all, but 🤷‍♂️ - it is quite similar to the problem with `AsyncTask`.

Now, it seems like (currently) it does interface with Java code (`NativeHttpRequest.java`). Actually, not sure how suspending functions look like from Java-code, but according to [stackoverflow answers](https://www.google.de/search?q=suspend+functions+in+java+code+site:stackoverflow.com&sca_esv=1c866aa5bb9857de&sa=X&ved=2ahUKEwjfmfm5-eKHAxVEwAIHHUW7F40QrQIoBHoECCMQBQ&biw=2048&bih=999&dpr=1.25), one can wrap it quite easily in a Java `CompletableFuture`.

Reply:
If did look at the same StackOverflow article, and came to the conclusion that you could **not** easily do this.  The most promising approach required 1) an extra library and 2) SDK version 24+.  I didn't want to commit a change which did that.

In any event, this code is only called from _NativeHttpRequest.java_, which is never used anywhere (hence the _@Keep_ annotation).

Reply:
Try this. Doesn't it work? It is included in kotlinx-coroutines-core. 

https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.future/

I think that `org.jetbrains.kotlinx:kotlinx-coroutines-jdk8` library you probably meant is obsolete or something, at least the current version of it is completely empty.

Reply:
Let me give that a try.  I'm not having any success getting that to compile.

Are you sure it's supported back to SDK 21?

Reply:
Hm no, on a second look, it appears that `CompletableFuture` in Android is only supported starting with SDK 24, after all:

https://developer.android.com/reference/java/util/concurrent/CompletableFuture

So it is not an issue with the coroutines library, but with Android.

Reply:
So, it looks like there is no easy way to interface suspending code to Android Java API 21+. 

Only when the code that uses this has also been migrated to Kotlin it could be made more Kotlin-idiomatic.